spyder/plugins/editor/extensions/docstring.py

```diff
@@ -0,0 +1,451 @@
+
+        indent1 = func_info.func_indent + self.code_editor.indent_chars
+        indent2 = func_info.func_indent + self.code_editor.indent_chars * 2
+
+        numpy_doc += '\n{}\n'.format(indent1)
+
+        if len(arg_names) > 0:
+            numpy_doc += '\n{0}Parameters'.format(indent1)
+            numpy_doc += '\n{}----------\n'.format(indent1)
+
+        arg_text = ''
+        for arg_name, arg_type, arg_value in zip(arg_names, arg_types,
+                                                 arg_values):
+            arg_text += '{}{} : '.format(indent1, arg_name)
+            if arg_type:
+                arg_text += '{}'.format(arg_type)
+            else:
+                arg_text += 'TYPE'
+
+            if arg_value:
+                arg_text += ', optional'
+
+            arg_text += '\n{}DESCRIPTION'.format(indent2)
+
+            if arg_value:
+                arg_value = arg_value.replace(self.quote3, self.quote3_other)
+                arg_text += ' (the default is {})'.format(arg_value)
+
+            arg_text += '\n'
+        numpy_doc += arg_text
+
+        numpy_doc += '\n{}Returns'.format(indent1)
+        numpy_doc += '\n{}-------'.format(indent1)
+        if func_info.return_type:
+            numpy_doc += '\n{}{}'.format(indent1, func_info.return_type)
+            numpy_doc += '\n{}DESCRIPTION\n'.format(indent2)
+        else:
+            numpy_doc += '\n{}RETURN_TYPE\n'.format(indent1)
+
+        numpy_doc += '\n{}{}'.format(indent1, self.quote3)
+
+        return numpy_doc
+
+
+class FunctionInfo:
+    """Parse function definition text."""
+
+    def __init__(self):
+        """."""
+        self.has_info = False
+        self.func_text = ''
+        self.args_text = ''
+        self.func_indent = ''
+        self.arg_name_list = []
+        self.arg_type_list = []
+        self.arg_value_list = []
+        self.return_type = None
+
+    @staticmethod
+    def is_char_in_pairs(pos_char, pairs):
+        """Return True if the charactor is in pairs of brackets or quotes."""
+        for pos_left, pos_right in pairs.items():
+            if pos_left < pos_char < pos_right:
+                return True
+
+        return False
+
+    @staticmethod
+    def find_quote_position(text):
+        """Return the start and end position of pairs of quotes."""
+        pos = {}
+        is_found_left_quote = False
+
+        for i, c in enumerate(text):
+            if is_found_left_quote is False:
+                if c == "'" or c == '"':
+                    is_found_left_quote = True
+                    quote = c
+                    left_pos = i
+            else:
+                if c == quote and text[i - 1] != '\\':
+                    pos[left_pos] = i
+                    is_found_left_quote = False
+
+        if is_found_left_quote:
+            raise IndexError("No matching close quote at: " + str(left_pos))
+
+        return pos
+
+    def find_bracket_position(self, text, bracket_left, bracket_right,

```

Comment:
You might be able to use this code to match brackets instead:
https://github.com/spyder-ide/spyder/blob/76edb5222ad8196791f6fc7177cf5eded1b6f542/spyder/plugins/editor/widgets/base.py#L449

Reply:
In order to use "find_brace_match", other parts of the code should also be modified.

Reply:
How difficult would that be to implement? That seems the one last issue holding up this PR, as it would greatly improve the usability and discoverability of this feature to have the automatic popup along with the existing shortcut/context menu command that users are unlikely to find on their own.

Reply:
@CAM-Gerlach Do you want to add UX2? or change "find_bracket_position" to "find_brace_match"?

Reply:
Oh sorry, maybe I was confused—I thought that was proposed as a way to work around UX2 relying on the CloseQuotes extension being enabled by the user.

Reply:
I will add the UX2 to this PR this weekend.

Reply:
Thanks! Looking forward to it!